### ISSUE TITLE
Proposal - Change Edit Tool Name

### ISSUE THREAD

**Issue by turtletooth** [opentoonz/opentoonz#1348]
The edit tool is actually more of an animate tool than an edit tool.  The selection tool is actually more of an edit tool.  
Would it be more accurate to call the edit tool the animate tool or something to that effect?

Just a thought.

**Comment by 2dvision** [comment]
I agree with "animate tools" name, But i think we need more merge Position, Rotation, Scale and Shear into one function (it is necessary). 

For example in krita:

![edit tools - merge function](https://user-images.githubusercontent.com/16509592/28736012-264312b6-73fd-11e7-8d67-2bb18e5bc487.gif)

**Comment by gab3d** [comment]
I support a name change, but to me "Transform Tool", as its name, would be more accurately reflecting what the tool does. Because either if the transformations would be animated or not would be an election of the user, but they will always be using it to actually transform an item's position, rotation, scale, etc.

**Comment by openanim** [comment]
I also thought of the name transform tool by seeing this issue as is the case in Toon boom harmony. But after reflection, in harmony the transform tool can animate only if the animate mode is activated. Otherwise, that can only transform. So, I think the name animate tool is better because that reflects the idea that you can animate with this tool.  The  selection tool in opentoonz looks more like the transform tool or select tool in harmony.

**Comment by turtletooth** [comment]
@gab3d I understand what you are saying, but I disagree.  Everything you do with the edit tool sets a keyframe - which is for animation.  The selection tool is also a transform tool - but without the keyframing.

**Comment by turtletooth** [comment]
This also makes for easier keyboard shortcut keys.  Right now eraser is "a" because "e" is taken for the edit tool.  

@2dvision Take a look at #1345 - the last icon in the row is a combo edit tool which does what you are hoping for.

**Comment by openanim** [comment]
@turtletooth Eventually even the selection tool does almost the same thing, it just missing a function shear.

**Comment by artisteacher** [comment]
I think the name change is a good idea. Edit as a name gives the wrong impression of the tool's purpose. If the name is changed, it will be important to have clear documentation. The harlequin manual is becoming less and less relevant as new features are merged.

**Comment by gab3d** [comment]
@turtletooth you know what? you're right, I'm not. :)

both are transform tools, there are two main differences between them:
1. Select Tool allows the transformation of elements inside a cell, whereas Edit Tool transforms the whole cell.
2. Edit Tool generates keyframes, whereas Select Tool does not.

So Transform Tool is not a good name.

**Comment by beeheemooth** [comment]
Maybe, "Action Tool"? 
It takes up less space than the "Animation Tool" and also starts on "a"

**Comment by RodneyBaker** [comment]
Edit Tool appears to be the most accurate thus far.
By selecting 'Edit' the basic properties of the object are presented to the user to change/edit.


As far as the word 'transform'...
Transform in 3D is the over arching term for Translation (movement in X,Y and Z axis), Scale and Rotation.  It is too limited to cover all of the options that are currently exposed via the Edit Tool or that some day might be available.  As an example of this last aspect, I don't think we can currently enter an expression as a value instead of an integer but if we could then the value of one property would then drive the other.  The point being... the object's property is 'edited'.  At present this might be animated or not (as has been stated) depending on whether there is a keyframe set to interpolate the data.  Transform would sometimes be an appropriate term but not always.  Edit, as of now, appears to always be an appropriate term.  Animate (not always as we may not be animating anything).  Action (a bit too vague again as we may just want to set some specific parameters).  

In other words, I think in essence when we select the Edit Tool we enter 'Edit Mode'.
Compare to 'Skeleton Mode', 'Paint Mode', 'Distortion Mode', etc. (which are not named thusly in OT) but the basic premise still stands.  

This does get me thinking about naming conventions in OT and it may be best not to fix what isn't broke.  For instance, all this talk has me thinking of the Rotate Tool.  The 'Rotate Tool' to me is a 'Turn Tool' as it doesn't actually rotate anything in the scene it just turns the canvas so that we get a different view of it.  To rotate anything we need to actually 'edit' the object's properties either with the Edit Tool or by entering data into the Function Editor.

I'll stop there before I hurt myself.  ;)

**Comment by turtletooth** [comment]
The main question should be "What is this tool's purpose?"  This tool is the primary way to animate properties in OpenToonz (although cut-out animators might argue that).  

Every tool edits something, so edit is about as non-descriptive as you can get.  I have, on more than one occasion, seen people trying to use the edit tool for simple transforms and changes to their images that should be done in the selection tool and then get confused why they are setting keyframes.

I think "edit" doesn't clearly explain - if you use this tool, things are going to start moving, while using "animate" makes it clear to all users that this tool is **primarily** used to make things move.

**Comment by Orphanlast** [comment]
I'm a Youtube Content creator that teaches people how to use Opentoonz, and I tend to agree with Turtletooth on most things Opentoonz Development related. I'm personally not apposed to the idea of calling the edit tool the "Animation Tool" Because that's exactly what it does. Calling it the "TRANSFORM TOOL" would be a mistake. Because it's not transforming anything. In fact, treating it as though it's transforming the image will run you into one mistake after another. Let's suppose your original image is rather large, And you "Transform" its size with the edit tool. Then you make a mesh.... the mesh will be the size of the original image before you "Transformed" it with the edit tool.

People complain to me about this on a regular basis. And I have to educate them that they're using the Edit tool incorrectly. 

I DO feel that calling it the "Animation Tool" would give newer users a better impression on what the tool actually does.

**Comment by snapai** [comment]
I definitely like the idea of this being renamed Animate Tool, but here are some other thoughts:

The idea of renaming the Edit tool to "Transform" exposes what is actually a feature "expectation" here - the idea that a column can have a static transform independent of its animation.

So if the tool was to be named Transform tool, then users would have the expectation of being able to do the transform operations without setting keyframes - e.g., by unchecking an "Animate" checkbox or something. The equivalent of using the Selection Tool set to "Whole Level" mode. (I was looking for the Animate checkbox myself when I first started using the Edit tool and was confused why it was setting keyframes).

I didn't realize that "Edit" was animation-only until I had accidentally animated a number of things with it by changing their size/position on different frames.

That carries over from 3D programs, where the Scene Object equivalent is a "transform" node that holds the transform data and rendering flags for the content, but not the mesh/spline data of the 3D object, regardless of whether it is animated or not. (Some programs, the animation simply overrides the transform the next time it is read, others the transform also transforms all the animation keys as well, causing the animation to occur at a different position, orientation, or size.)

However, as the Edit tool currently operates, "Animate" best describes what it does - you should expect a keyframe to be set if you change something about a Column using this tool.

**Comment by RodneyBaker** [comment]
I'm starting to sway... but with some serious caveats.  ;)
For the sake of discussion, let's say the Edit Tool is Renamed 'Animate' or 'Animation' of which 'Animate is likely preferred due to length, etc.  In other words, if we are going to fix this... perhaps we should fix (as in repair and set) all of the tool names.

This change should likely drive similar changes in the other 'tools' for brevity and consistency's sake.
For instance:
**Edit Tool > Animate (Shortcut change to A with Erase shortcut change to E)**
Selection Tool > Select (Shortcut S)
Brush Tool > Brush (Shortcut B)
Geometric Tool > Geometry (Shortcut G)
etc.
At this point further changes in other Tool names and shortcuts becomes largely a matter of personal preference so short of dropping the word 'Tool' I'd say they would be reserved as separate PRs.  For instance as I've mentioned before the Rotate Tool would more ideally be 'Turn' (see prior explaination for that change).  This would likely create a conflict with the Type Tool's shortcut of 'T' if altered from it's current Shift + Spacebar.  But... in short, there may be little need to state 'Tool' in the tooltip/documentation.

Aside:  While I can see the value in an accompanying shortcut key change, care should be taken in changing of shortcuts especially with a swap of A and E because in thinking we are about to animate something older users of OT will be set to erase it.  This might also have a longer tail in code repair if such shortcuts are exposed for use with scripting.

So I see three related PRs potentially emanating from this; this PR being to change the name and shortcuts for Edit Tool and by extension the Erase Tool.  The second would be to address the low hanging fruit of (mostly) superficial changes such as dropping the word 'Tool' and the third being to target resolution of the remainder of tool names and shortcuts.

**Comment by RodneyBaker** [comment]
I'd like to address the following separately because I think it has important workflow potential with regard to the Edit/Animate Tool...

@snapai said:

> as the Edit tool currently operates, "Animate" best describes what it does - you should expect a keyframe to be set if you change something about a Column using this tool.

If not already present, I believe there should be a means to deactivate the automatic setting of keyframes while using Edit/Animate,  Currently a keyframe appears to be generated with any use of the Edit Tool but it should be possible bypass the setting of keyframes while using the tool.   In other software I use this is controlled by toggling on/off an Animate mode with the default being set to animate (set keyframes). 

The UI in the case I am thinking even places a red border around the viewing screen (optionally reversible) that visually lets the user know they will be setting keyframes.  If toggled off, the user can manipulate objects in the scene without concern for setting keyframes anywhere on the timeline/xsheet.  If toggled on a keyframe will be set according to additional options set for keyframing.  In this case a separate toggle being available for keyframing one or more of the transforms (translate, rotate, scale, mesh, etc.) or toggle an icon that activates the keyframing of 'all'.  ( I may need to demonstrate this but this level of funcxtionality allows for some very fine differentiation in the setting of keyframes according to user set parameters.  If they only want to animate rotation in a scene they can exclude any inadvertent animation of other settings simply by toggling off all options other than rotation.

All of this delves into the realm of keyframing which is a bit beyond the scope of this PR but it is related in that the Edit Tool/Animate Tool is not exclusively about animating with keyframes.

**Perhaps... as a middleground that allows this PR to be implemented the tool could be renamed to 'Edit/Animate (A)' or 'Animate/Edit (A)'... of which I would prefer the former as it suggests exactly what the tool does.  This will be a little longer than some would prefer but I believe would would resolve the underlying premise behind the PR.**  If desired, a second PR might be initiated to drop the word 'Tool' from the other tooltips.

**Comment by RodneyBaker** [comment]
I will add that a change in name to 'Edit/Animate (A)' also orients the tool in the direction that @2dvision suggests of further integrating the various transform tools at the user interface.  I have some reservation with regard to any superficial adoption of combined transforms but the suggestion hints at deeper functionality that can and should be addressed in the future with regard to keyframing.

**Comment by blurymind** [comment]
@turtletooth here is a request that I made a while ago:
https://github.com/opentoonz/opentoonz/issues/157
I think that @2dvision  is right to ask for it - a multi-task manipulator is available in toonz harlequin, 
![c63faa02-4ee9-11e6-9134-1152faab61aa](https://user-images.githubusercontent.com/6495061/28768092-4bdf648e-75ce-11e7-9a38-04dcca223d83.gif)

however for what ever reason it was removed in OT and now you have to change between modes when animating. I strongly believe that greatly reduces productivity

I understand @shun-iwasawa 's concerns, but please at least add it back as an optional mode:
![e6129806-5067-11e6-8d06-d176f4252220](https://user-images.githubusercontent.com/6495061/28768141-70c77552-75ce-11e7-9b69-0964fbefa0f6.png)

**Comment by openanim** [comment]
The name animate tool is the most appropriate on what edit tool does. Harmony has an animate mode that allows the creation of keyframe when enabled. The edit tool appears to work in the same way as the harmony transformation tool in animation mode enabled.

Another name that might fit would be Keyframer tool like the fx available in tvpaint in the animation menu. That creates animation key over time.

The idea that we create an animation over time depending on how we move objects should be easier to understand. A bubble info as suggested could help.

@blurymind this gadget is awesome in harlequin, I think it is possible to regain its functions thanks to the super work in preparation mentioned here #1345

**Comment by gab3d** [comment]
my position about this issue is:

- in the current state of OT I think "Animate" is the best name sugested for the tool.

- if there would exist an Animation Mode in OT (which to me would be the best case) then "Transform" would be a better name for the Edit tool.

The other thing that bothers me is the name of the "Select" tool, because it does so much more than just allow for the selection of things, instead it allows to transform things too.
So... the whole issue turns put a bit more complicated to address.
In my mind neither Select nor Edit represent good (or standard) name for these tools and what they do.
Considered independently both nanes are not bad, but when you consider them together... well, both of them edit things, select things, even transform things.
Animate differentiate Edit tool from Select tool well in the current OT workflow, but even then I think an Animation Mode would bring a more standard workflow into OT, and then Animate would just not cut it as the name of that tool either.
Perhaps the type of element they act upon could be considered as a differential. Objects vs. ??? (don't know how image elements are called inside OT)

Bottom line is... I still don't have a perfect candidate. :-/

**Comment by blurymind** [comment]
@openanim I really hope that @agnyy manages to merge his fix,which seemsto include a universal transform widget that contains all the operations in a single place- much like the one available in harlequin.

The one we have atm in OT is a mere crippled version of the one in toonz harlequin- that adds the need to do laborious boring clicking in drop list menu between moving,rotating and scaling.
It almost feels like punishment to use it :(

**Comment by turtletooth** [comment]
Just found another old example of the confusion in #859

**Comment by turtletooth** [comment]
@shun-iwasawa What are your thoughts on this?

**Comment by shun-iwasawa** [comment]
Sorry for my late response and thank you for deep discussions done so far.

I understand @gab3d suggested the "Transform" from the developer's point of view , since what this tool actually does is to edit the "transformation matrix" of the current column. But unfortunately it is also true that users are quite unfamiliar with such word.

For the "Animate tool", I understand it'll describe some aspect of this tool (especially, focusing on that this tool creates keyframe-based movement). But in my very-biased opinion from Japanese animation industry, to "animate" (= bring to life) something can be achieved by sequential drawings drawn by animator, rather than by just translating, scaling or rotating.

Anyway, it's OK to change it to "Animate" - since it seems that the majority support the "Animate tool" in the discussion above. I think it's just a matter of getting used to it.

**Comment by konero** [comment]
I share the same sentiment as @shun-iwasawa, but I was trained in the Japanese industry. For a traditional animator, "Animate Tool" won't make sense to me. Everything the tool does I would probably term "staging" when editing levels or "shooting" when editing a camera.

Similar deal with stop-motion, by the time all the frames have been shot the animating part is over.

**Comment by gab3d** [comment]
@shun-iwasawa, I really sugested "Transform" because is the term usually used in several applications (like Krita, Photoshop, and 3D animation soft) to encompass all of translation/rotation/scale operations.
For example in Photoshop, when you go to the Edit menu, you can find the Free Transform tool (Ctrl+t), which gives a manipulator to change this three characteristics of a layer (much like the Selection Tool in OT, which name is also confusing, by the way, as it does so much more than selecting things).

![photoshop-free-transform-command](https://user-images.githubusercontent.com/18104091/29695922-e64c4ad6-891b-11e7-9e06-bec17b108c29.png)
...or in Krita:
![krita](https://user-images.githubusercontent.com/18104091/29695539-ec131762-8919-11e7-8ff5-899fcaf32606.png)
![blender](https://user-images.githubusercontent.com/18104091/29696672-ce720c16-8920-11e7-9aea-b4537121a563.png)
![maya](https://user-images.githubusercontent.com/18104091/29696677-d2d7596e-8920-11e7-8838-e2d55cc5fc4e.png)


To me the ideal situation would be that there was an Animation Mode button in OT, permanently sitting somewhere in the UI, and that while that mode was ON both tools (current Edit and Selection tools) would animate the items they respectively work on (layers and strokes/selected area).
With this approach in mind, probably the best name for these tools would be "Transform Layer" and "Transform Selection" (or something like that).

These tools would immediately animate layers or strokes, if the Animation Mode were ON, 
and just transform them (without generating keys) if that mode was OFF.
Conveniently enough, that would allow the Transform Layer and Transform Selection tools to be used to offset the position/rotation/scale values of entire animations, provided keyframes did exist for that items and the Animation Mode was OFF.

This scheme could also be applicated to the current "Control Point Editor" tool, which could then:
1. be renamed to "Transform Control Points" for coherence.
2. allow to transform (move/rotate/scale) a group of control points, using a similar manipulator as the one used by the current Selection (then Transform Selection) tool.
3. allow to animate strokes' control points positions, when the Animate Mode was ON (and perhaps offset them when Animation Mode is OFF).

That, in my mind, would be the ideal workflow, and the most clear naming convention for these tools also. 

Having said that, I understand that these changes would not be easy to implement or happen soon, so I could live with any other tool names you come up with.  ;)

**Comment by snapai** [comment]
Sorry to bring up an old discussion again, but one more thing I discovered:
Some of the FX schematic nodes have widgets for their values, adjustable on the canvas only while using the "edit" tool.
HOWEVER, using that widget to adjust their values also keyframes them - while typing in the values into the node's parameters dialog does not.

There's definitely a need for an "Auto-keyframe" or "Animate" mode toggle in the UI, it's going to be a common user expectation, particularly since all other software has it. This is usually either a button with a "key" on it, or in apps that use a key button to set keyframes manually (like Blender), the "red dot" of a recording camera to indicate changes will be recorded.

If that's created then it's a matter of making sure that the data model reacts correctly to edits made with edit mode off when keyframe animation already exists for the channels being edited.

There's a few different options for handling that:

- The edits are displayed, but not stored unless manually keyed, so that any change to the current time reloads the values from the animation curves.
- Each property has an un-animatable "base value" that the animation is added to/multiplied with: adjusting the value with auto-keyframe off offsets the entire animation.
- The value is locked if keyframed, and unable to be changed except in the dope sheet/graph editor/function editor

The first is useful when combined with a workflow that allows disabling updates when changing the current time. You can move to a new time without loading the animation, adjust the position, then set a key on the values you want to keep - as if there were no keyframes previously set. Good for stopmotion-esque animation as well as an alternate way to copy keyframes from one frame to another.

The second provides a different layer of flexibility: an animation property can be adjusted as a whole without creating a new layer or subxsheet to adjust it with. However it requires the most code overhead for a nearly "invisible" improvement - making sure that every property that can be animated also has an adjustable but not animatable base value. Since this can also be done via layer parenting or sub-xsheets, or manually shifting an entire curve in the graph editor, it might not be appropriate to add this capability a third time here, despite being quite useful

The third is possibly the easiest to implement - simply passing the facts onto the user: if you're not creating a new keyframe on an animated value, this value is not going to change. However it will break user expectations and create a lot of questions and complaints later without an error dialog saying "This property is already controlled by an animation curve, enable animation mode to edit it" - otherwise it appears that a tool in the UI simply stopped working.

**Comment by shun-iwasawa** [comment]
@snapai 

> HOWEVER, using that widget to adjust their values also keyframes them - while typing in the values into the node's parameters dialog does not.

Pardon me if I'm misunderstanding your comment, but you can set key by clicking the "key" button <img src="https://user-images.githubusercontent.com/17974955/31064817-f8b30020-a77a-11e7-961a-58f8fc5f0c48.png">  located next to each parameter in the Fx settings dialog.
It has 4 states as follows:

- It is grey & unpressed if no key value is defined for the parameter at the current frame.
- It is orange & pressed when a key value is defined for the parameter at the current frame.
- It is grey & pressed if the parameter is animated but no key value is defined at the current frame.
- It is yellow & pressed if you change the parameter value and the current frame does not have a key for the parameter.

Is it enough for you? You meant that you would like to have an option to set keys as soon as you change the parameter value without clicking the key button (i.e. skipping the "yellow" state of the key button)?

**Comment by gab3d** [comment]
I think he's just speculating on what would be the expected bahavior if an Animation Mode (i.e. Auto Key button) existed in OT, and it was activated.

**Comment by gab3d** [comment]
Speaking of all this issues, I think OT in the future could have two new UI elements:

1. Dockable bar including:
    a. Combo box for showing (and changing) currently selected stage object (col, camera, peg).
    b. "Animation Mode" toggle button. (as discussed in several posts above)

2. A general (dockable) Properties panel, where the current frame values of the properties of the selected element could be shown.
That is, if the selected element is a Column, its current transform values would be displayed and edited too (pos, rot, scale, shear, SO, opacity).
If it's a Level, their properties would be shown.
If it's a Node, then the node properties would be exposed there. For example, for a FX node, the FX properties qould be shown. For the Xsheet node, for example, the Scene properties could be shown. For a Camera node, the camera settings would be shown. For an Output node, the Output Options shown and so on...

It would function in very much the same way as the properties panel in Nuke or Natron, or the Atrribute Editor in Maya, to name some references.

That would alleviate the need for space in the tools optionss panel too, as it would take out all the N/S, E/W, Z, SO stuff off there.

I know this would take a considerable rethinking of the UI concept, and goes far beyond the scope of the current issue here, so my intention is just to bring an idea to the table for future discussion.

**Comment by snapai** [comment]
@shun-iwasawa Sorry no, I was not making a request there, I was describing the program's current behavior.
![image](https://user-images.githubusercontent.com/5098372/31213063-56cfdffe-a972-11e7-9b98-491b33ac0e1f.png)
The values are currently not animated and no keys are set.
If I adjust these values by typing in new values, the values are changed, no key is set.
If I adjust them by using the "edit tool" on the UI widget, the values are changed, an animation key is inserted.
If the key is removed, they return to their previous value.
The UI widget does not appear unless the "Edit" tool is selected - therefore there is no way to use the widget to adjust the value without inserting keyframes.

**Comment by shun-iwasawa** [comment]
@snapai Understood, thank you for the response. Your description is correct.
